### ISSUE TITLE
Labels and overlays

### ISSUE THREAD

**Issue by vascotenner** [ioam/holoviews#110]
In ipython notebook, I try to display titles on top of an overlay. If I put in a cell:
```python
%opts Overlay [show_title=True]
arrows_overlay = hv.Arrow(0.4,0.4, '', direction='^', group='Far')
(Image(far, group='Far')*arrows_overlay)
```
This will show a title on top.

I set the opts with %opts, and I would expect that all subsequent overlays will produce the same result. However, when I execute the same cell with:
```python
(Image(far, group='Far')*arrows_overlay)
```

I do not see a title!

**Comment by jlstevens** [comment]
Are you using the latest development version of holoviews? What is the version reported with ``str(hv.__version__)``? I am asking as this example works fine for me:

```python
%opts Overlay [show_title=True]
import holoviews as hv
import numpy as np
arrows_overlay = hv.Arrow(0.4,0.4, '', direction='^', group='Far')
far = np.random.rand(10,10)
hv.Image(far, group='Far') * arrows_overlay
```

Then this works as expected:

```python
hv.Image(far, group='Far')*arrows_overlay
```

We are due to release on PyPI in the next two days so if you aren't using the development version, you might want to wait.

**Comment by vascotenner** [comment]
I am using '0.8.1-1693-g63ae68c'

In my notebook I executed some cells before. This issue might be related to: https://github.com/ioam/holoviews/issues/93

Here the  code that generates problems:
```%opts Overlay.Far [show_title=False]```

Next cell (generates plot with title):
```
%opts Overlay [show_title=True]
import holoviews as hv
import numpy as np
arrows_overlay = hv.Arrow(0.4,0.4, '', direction='^', group='Far')
far = np.random.rand(10,10)
hv.Image(far, group='Far') * arrows_overlay
```

Next cell (generates plot **without** title):
```hv.Image(far, group='Far') * arrows_overlay```

**Comment by jlstevens** [comment]
That version string ( '0.8.1-1693-g63ae68c') is wrong but it should be fixed if you run ``git pull --tags``. Can you do that and try that again please?

Looking at your code, I think I see the problem:

```python
%opts Overlay.Far [show_title=False]
```

This will override any options for Overlays *if* the group is set to 'Far' because it is a more specific specification.  You either need to:

* Create two group names, one with titles and one without and the corresponding ``%opts`` (line magic) definition.
* Use the ``%%opts`` cell magic to disable the title in one specific cell (recommended)
* There is a third way using ``__call__`` but we are currently fixing a bug in that right now (arrow is causing trouble). I will post this approach once we have it working.

Generally, it is recommended that groups are groups of options that are semantically useful i.e define a group for a collection of things that need to be displayed a certain way because of what the data in them represents.  

For displaying something is a particular cell in a particular way just for that cell, we would recommend using ``%%opts``.

**Comment by jlstevens** [comment]
I have just pushed commit c84fbc2 which fixes a problem when using ``__call__`` on overlays involving ``Arrow`` or ``Text`` elements. Now you can use the third way of customizing the options to disable the title:

* Use ``__call__` on your overlay when you want to disable the title:

```python
(hv.Image(far, group='Far') * arrows_overlay)(plot=dict(show_title=False))
```

I think I will close the issue now as everything is working as expected. That said, I would be happy to hear any suggestions if you have any suggestions/proposals to make the options system more useful or easier to understand!

**Comment by vascotenner** [comment]
The problem I wanted to address is that both plots are different, while I only set "GLOBAL" options via %opts.

**Comment by jlstevens** [comment]
I think there may be some confusion about what is meant by 'global'.

The line magic applies throughout the notebook session (as opposed to the cell magic which applies only for one cell). This is the only difference between ``%opts`` and ``%%opts`` - the former will apply between cells while the former doesn't.

This meaning of 'global' doesn't mean that the ``%opts`` line magic definition will affect all ``Overlays`` (in the example) in the same way: if you have a definition using a group (``'Far'`` in this case) then if what you define in ``%opts`` is made specific to ``Far`` (i.e where you did ``%opts Overlay.Far [show_title=False]``) then only ``Overlay`` objects with that particular group specified will be affected.

Hope that makes sense and we will have to think about how we can make this clearer in the documentation.

**Comment by vascotenner** [comment]
I think I did not address the problem properly:
After executing an %opts statement, I suspect that all the plots in the rest of the session are looking the same. However, in this it does depend if make the plot in the same cell as the %opts, or in the next cell. This behaviour is only expected for %%opts.

**Comment by jbednar** [comment]
Maybe you could provide another example, and explain precisely what it shows and what you had hoped it to do?  I can't follow the explanation in your last comment.

**Comment by vascotenner** [comment]
I expect that both plots are identical. This is what is going wrong:
![screenshot](https://cloud.githubusercontent.com/assets/3776190/7861791/83e30a5a-0552-11e5-810e-b9c40209bee8.png)

Version: '1.1.0-324-g5486c1e'

**Comment by philippjfr** [comment]
Thanks for reporting that. There was a bug in the OptsMagic, which I've now fixed in 00b03dbdd. Looks like we'll re-release v1.2, because that's a pretty important bug.

**Comment by jlstevens** [comment]
Thanks for supplying the example, it helped me understand what exactly was going wrong.

I've just fixed the PyPI release to include Philipp's commit. Hopefully we won't find any other urgent fixes otherwise we will have to release 1.2.1 in short order!
